odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2372 @@
+    def device(self):
+        """The GPU device on which this tensor lies."""
+        return self.space.device
+
+    def asarray(self, out=None):
+        """Extract the data of this element as a `numpy.ndarray`.
+
+        Parameters
+        ----------
+        out : `numpy.ndarray`, optional
+            Array to which the result should be written.
+            Has to be contiguous and of the correct data type.
+
+        Returns
+        -------
+        asarray : `numpy.ndarray`
+            Numpy array of the same `dtype` and `shape` this tensor.
+            If ``out`` was given, the returned object is a reference to it.
+
+        Examples
+        --------
+        By default, a new array is created:
+
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> x.asarray()
+        array([ 1.,  2.,  3.])
+        >>> int_spc = odl.tensor_space(3, impl='cupy', dtype=int)
+        >>> x = int_spc.element([1, 2, 3])
+        >>> x.asarray()
+        array([1, 2, 3])
+        >>> tensors = odl.rn((2, 3), impl='cupy', dtype='float32')
+        >>> x = tensors.element([[1, 2, 3],
+        ...                      [4, 5, 6]])
+        >>> x.asarray()
+        array([[ 1.,  2.,  3.],
+               [ 4.,  5.,  6.]], dtype=float32)
+
+        Using the out parameter, the array can be filled in-place:
+
+        >>> out = np.empty((2, 3), dtype='float32')
+        >>> result = x.asarray(out=out)
+        >>> out
+        array([[ 1.,  2.,  3.],
+               [ 4.,  5.,  6.]], dtype=float32)
+        >>> result is out
+        True
+        """
+        if out is None:
+            return cupy.asnumpy(self.data)
+        else:
+            if out.shape != self.shape:
+                raise ValueError('`out` must have shape {}, got shape {}'
+                                 ''.format(self.shape, out.shape))
+            if out.dtype != self.dtype:
+                raise ValueError('`out` must have dtype {}, got dtype {}'
+                                 ''.format(self.dtype, out.dtype))
+            self.data.data.copy_to_host(
+                out.ctypes.data_as(np.ctypeslib.ctypes.c_void_p),
+                self.size * self.itemsize)
+            return out
+
+    @property
+    def data_ptr(self):
+        """A raw pointer to the data container.
+
+        Examples
+        --------
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.one()
+        >>> x.data_ptr  # doctest: +SKIP
+        47259975936
+        """
+        return self.data.ptr
+
+    def __eq__(self, other):
+        """Return ``self == other``.
+
+        Parameters
+        ----------
+        other :
+            Object to be compared with ``self``.
+
+        Returns
+        -------
+        equals : bool
+            ``True`` if all entries of ``other`` are equal to this
+            tensor's entries, ``False`` otherwise.
+
+        Notes
+        -----
+        The element-by-element comparison is performed on the CPU,
```

Comment:
Really? Why not write a `ReductionKernel` for this? Should be rather simple

Reply:
Makes sense.